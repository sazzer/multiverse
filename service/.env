RUST_LOG=multiverse_lib=debug,info
DATABASE_URL=postgres://multiverse:multiverse@localhost:45432/multiverse