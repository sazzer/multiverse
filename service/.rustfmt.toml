max_width = 100
edition = "2018"
merge_imports = true
normalize_comments = true
normalize_doc_attributes = true
reorder_impl_items = true
use_field_init_shorthand = true
