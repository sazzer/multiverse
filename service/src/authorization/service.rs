use chrono::Duration;

mod generate_authorization;

/// The Authorization Service to work with
#[derive(Clone)]
pub struct AuthorizationService {
    /// The duration of time that a generated authorization is valid for
    duration: Duration,
}

impl AuthorizationService {
    /// Create a new instance of the Authorization Service
    ///
    /// # Parameters
    /// - `duration` - The duration of time that a generated authorization is valid for
    ///
    /// # Returns
    /// The Authorization Service ready to use
    pub(super) fn new(duration: Duration) -> Self {
        AuthorizationService { duration }
    }
}
