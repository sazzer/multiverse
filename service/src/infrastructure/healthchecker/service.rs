use super::check_health::CheckHealth;
use super::model::{ComponentHealth, SystemHealth};
use std::{collections::HashMap, sync::Arc};

/// System to check the health of the application
pub struct Healthchecker {
    /// The set of components whos health is to be checked
    components: HashMap<String, Arc<dyn CheckHealth>>,
}

impl Healthchecker {
    /// Construct a new healthchecker to check the health of the system
    ///
    /// # Parameters
    /// - `components` - The components to check as part of the system
    ///
    /// # Returns
    /// The newly constructed health checker to work with
    pub fn new(components: HashMap<String, Arc<dyn CheckHealth>>) -> Self {
        Self { components }
    }

    /// Check the health of the entire system.
    ///
    /// This calls each of the individual healthchecks that are registered and uses the results from those
    /// to compile the overall health of the system
    ///
    /// # Returns
    /// The health of the system as a whole
    pub async fn check_health(&self) -> SystemHealth {
        let mut result = HashMap::new();

        for (name, component) in &self.components {
            let component_health = component.check_health().await;
            tracing::debug!(component = ?name, health = ?component_health, "Component health");
            result.insert(
                name.clone(),
                match component_health {
                    Ok(_) => ComponentHealth::Healthy,
                    Err(e) => ComponentHealth::Unhealthy(format!("{}", e)),
                },
            );
        }

        SystemHealth::new(result)
    }
}

/// Builder to allow easier construction of a `Healthchecker` by allowing components to be added individually
#[derive(Default, Clone)]
pub struct HealthcheckerBuilder {
    /// The set of components whos health is to be checked
    components: HashMap<String, Arc<dyn CheckHealth>>,
}

impl HealthcheckerBuilder {
    /// Add a new component to the health checker that we are building
    ///
    /// # Parameters
    /// - `name` - The name of the component to add
    /// - `component` - The actual component to add
    ///
    /// # Returns
    /// The builder, so that this call can be chained
    ///
    /// #Type Parameters
    /// - `<S>` - Anything that can convert into a `String`, to allow for easier calling
    pub fn with_component<S>(&mut self, name: S, component: Arc<dyn CheckHealth>) -> &mut Self
    where
        S: Into<String>,
    {
        self.components.insert(name.into(), component);

        self
    }

    /// Actually build the health checker
    ///
    /// # Returns
    /// A Healthchecker instance that represents the same set of components
    pub fn build(self) -> Healthchecker {
        Healthchecker::new(self.components)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[actix_rt::test]
    async fn test_no_components() {
        let builder = HealthcheckerBuilder::default();
        // builder.with_component("passing", Arc::new(Ok(())));
        let sut = builder.build();

        let result = sut.check_health().await;
        assert_eq!(true, result.is_healthy());
        assert_eq!(true, result.components.is_empty());
    }

    #[actix_rt::test]
    async fn test_passing_components() {
        let mut builder = HealthcheckerBuilder::default();
        builder.with_component("passing", Arc::new(Ok(())));
        let sut = builder.build();

        let result = sut.check_health().await;
        assert_eq!(true, result.is_healthy());
        assert_eq!(1, result.components.len());
        assert_eq!(
            Some(&ComponentHealth::Healthy),
            result.components.get("passing")
        );
    }

    #[actix_rt::test]
    async fn test_failing_components() {
        let mut builder = HealthcheckerBuilder::default();
        builder.with_component("failing", Arc::new(Err("Oops")));
        let sut = builder.build();

        let result = sut.check_health().await;
        assert_eq!(false, result.is_healthy());
        assert_eq!(1, result.components.len());
        assert_eq!(
            Some(&ComponentHealth::Unhealthy("Oops".to_owned())),
            result.components.get("failing")
        );
    }

    #[actix_rt::test]
    async fn test_mixed_components() {
        let mut builder = HealthcheckerBuilder::default();
        builder.with_component("passing", Arc::new(Ok(())));
        builder.with_component("failing", Arc::new(Err("Oops")));
        let sut = builder.build();

        let result = sut.check_health().await;
        assert_eq!(false, result.is_healthy());
        assert_eq!(2, result.components.len());
        assert_eq!(
            Some(&ComponentHealth::Healthy),
            result.components.get("passing")
        );
        assert_eq!(
            Some(&ComponentHealth::Unhealthy("Oops".to_owned())),
            result.components.get("failing")
        );
    }
}
