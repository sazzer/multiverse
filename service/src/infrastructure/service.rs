use super::server::{testing::TestResponse, Server};
use crate::{
    authentication::configure::AuthenticationConfig,
    infrastructure::{database, healthchecker::configure::HealthcheckConfig},
    users::configure::UsersConfig,
};
use std::sync::Arc;

/// The actual service that represents the entire system.
pub struct Service {
    server: Server,
}

impl Service {
    /// Construct a new service, configured with the provided settings and ready to run
    ///
    /// # Parameters
    /// - `settings` - The settings needed to build the application
    ///
    /// # Returns
    /// The built service, ready to work with
    pub async fn new(settings: crate::Settings) -> Self {
        tracing::info!(settings = ?settings, "Building service");

        let database = database::Database::new(settings.database_url).await;
        database::migrate::migrate_database(&database)
            .await
            .expect("Failed to migrate database");

        let users = UsersConfig::new(database.clone());
        let authentication = AuthenticationConfig::new();

        let healthchecks = HealthcheckConfig::default().with_component("db", Arc::new(database));

        Service {
            server: Server::new(vec![
                healthchecks.configure(),
                users.configure(),
                authentication.configure(),
            ]),
        }
    }

    /// Start the service listening on the given HTTP port and accepting incoming connections
    ///
    /// # Parameters
    /// - `port` - The port to listen on
    pub async fn start(&self, port: u16) {
        tracing::info!(port = port, "Starting service");
        self.server.start(port).await;
    }

    /// Send an HTTP Request in to the service and return the response
    ///
    /// # Parameters
    /// - `request` - The request to send to the service
    ///
    /// # Returns
    /// The HTTP Response
    pub async fn test_request(&self, request: actix_http::Request) -> TestResponse {
        self.server.test_request(request).await
    }
}
