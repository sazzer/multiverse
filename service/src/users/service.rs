mod create_user;
mod find;
mod update;

use super::repository::UserRepository;
pub use create_user::CreateUserError;

/// Service Layer for dealing with Users
#[derive(Clone)]
pub struct UsersService {
    repository: UserRepository,
}

impl UsersService {
    /// Create a new instance of the Users Service
    ///
    /// # Returns
    /// The Users Service ready to use
    pub(super) fn new(repository: UserRepository) -> Self {
        Self { repository }
    }
}
