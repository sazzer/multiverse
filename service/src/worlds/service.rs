mod create_world;
mod find;
use super::repository::WorldRepository;
pub use create_world::CreateWorldError;

/// Service Layer for dealing with Worlds
#[derive(Clone)]
pub struct WorldsService {
    repository: WorldRepository,
}

impl WorldsService {
    /// Create a new instance of the Worlds Service
    ///
    /// # Returns
    /// The Worlds Service ready to use
    pub(super) fn new(repository: WorldRepository) -> Self {
        Self { repository }
    }
}
