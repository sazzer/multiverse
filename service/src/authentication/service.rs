use crate::users::UsersService;

/// Service for authenticating a user
#[derive(Clone)]
pub struct AuthenticationService {
    /// The Users Service, to create and retrieve user records
    _users_service: UsersService,
}

impl AuthenticationService {
    /// Create a new instance of the Authentication Service
    ///
    /// # Parameters
    /// - `users_service` - The users service
    ///
    /// # Returns
    /// The Authentication Service ready to use
    pub fn new(users_service: UsersService) -> Self {
        AuthenticationService {
            _users_service: users_service,
        }
    }
}
