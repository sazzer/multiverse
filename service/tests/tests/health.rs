use crate::service::TestService;
use insta::{assert_debug_snapshot, assert_json_snapshot};

#[actix_rt::test]
async fn integration_test_healthcheck() {
    let service = TestService::new().await;
    let response = service.request(actix_web::test::TestRequest::get().uri("/health").to_request()).await;

    assert_debug_snapshot!("healthcheck-headers", response.headers());
    assert_json_snapshot!("healthcheck-body", response.to_json().unwrap());
}
