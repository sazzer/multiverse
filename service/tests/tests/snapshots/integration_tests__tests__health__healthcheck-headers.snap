---
source: tests/tests/health.rs
expression: response.headers()
---
"HTTP/1.1 200 OK\ncontent-type: application/json"
